app/web_framework/cluster_slave_application.py

```diff
@@ -116,6 +126,82 @@ def get(self, build_id, subjob_id):
         self.write(response)
 
 
+class _AtomsHandler(_ClusterSlaveBaseAPIHandler):
+    def get(self, build_id, subjob_id):
+        response = {
+            'comment': 'not implemented',
+            }
+        self.write(response)
+
+
+class _AtomHandler(_ClusterSlaveBaseAPIHandler):
+    def get(self, build_id, subjob_id, atom_id):
+        response = {
+            'comment': 'not implemented',
+            }
+        self.write(response)
+
+
+class _AtomConsoleHandler(_ClusterSlaveBaseAPIHandler):
+    def get(self, build_id, subjob_id, atom_id, max_lines=50, offset_line=None):
+        """
+        Return the console output if it exists, returns a 404 otherwise.
+
+        On success, the response contains keys: offset_line, num_lines, total_num_lines, and content.
+
+        e.g.:
+        {
+            'offset_line': 0,
+            'num_lines': 50,
+            'total_num_lines': 167,
+            'content': 'Lorem ipsum dolor sit amet,\nconsectetur adipiscing elit,\n...',
+        }
+
+        :type build_id: int
+        :type subjob_id: int
+        :type atom_id: int
+        :param max_lines: The maximum total number of lines to return. If this max_lines + offset_line lines do not
+            exist in the output file, just return what there is.
+        :type max_lines: int
+        :param offset_line: The line number (0-indexed) to start reading content for. If none is specified, we will
+            return the console output starting from the end of the file.
+        :type offset_line: int | None
+        """
+        if offset_line is not None and offset_line < 0:
+            response = {'message': '\'offset_line\' must be greater than or equal to zero.'}
+            self._write_status(response, success=False, status_code=400)
+            return
+        if max_lines <= 0:
+            response = {'message': '\'max_lines\' must be greater than zero.'}
+            self._write_status(response, success=False, status_code=400)
+            return
+
+        artifact_dir = BuildArtifact.atom_artifact_directory(build_id, subjob_id, atom_id)
+        output_file = os.path.join(artifact_dir, BuildArtifact.OUTPUT_FILE)
```

Comment:
How do you feel about moving most of this logic in this method into `ClusterSlave`? Originally the `ClusterMasterApplication` and `ClusterSlaveApplication` classes were intended to just be thin wrappers around `ClusterMaster` and `ClusterSlave`. (The original motivation was that we should ideally be able to swap out the web framework layer without affecting underlying app behavior.)

We've moved away from that a bit and some business logic has crept in to the web framework layer, but I think it's still a good ideal to shoot for.

Something you could also take advantage of is from `ClusterSlave` raise a `BadRequestError` with the error message and the web framework layer will automatically create the `{'message': 'whatever'}` response instead of you needing to construct it manually here.